android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIText.java

```diff
@@ -707,50 +711,196 @@ public void handleKeyboard(KrollDict d)
-			}
-
-			tv.setCursorVisible(true);
+			boolean customKeyListener = false;
 			switch (type) {
 				case KEYBOARD_DEFAULT:
 				case KEYBOARD_ASCII:
 					// Don't need a key listener, inputType handles that.
 					break;
 				case KEYBOARD_NUMBERS_PUNCTUATION:
-					textTypeAndClass |= (InputType.TYPE_CLASS_NUMBER | InputType.TYPE_CLASS_TEXT);
-					tv.setKeyListener(new NumberKeyListener() {
-						@Override
-						public int getInputType()
-						{
-							return InputType.TYPE_CLASS_NUMBER | InputType.TYPE_CLASS_TEXT;
-						}
-
-						@Override
-						protected char[] getAcceptedChars()
-						{
-							return new char[] { '0', '1', '2', '3', '4', '5', '6', '7',  '8', '9', '.',
-												'-', '+', '_', '*', '-', '!', '@', '#',  '$', '%', '^',
-												'&', '*', '(', ')', '=', '{', '}', '[',  ']', '|', '\\',
-												'<', '>', ',', '?', '/', ':', ';', '\'', '"', '~' };
-						}
-					});
+					final char[] numbersAndPunctuationChars =
+						new char[] { '0', '1', '2',  '3', '4', '5', '6', '7', '8', '9', '.',  '-', '+', '_', '*',
+									 '-', '!', '@',  '#', '$', '%', '^', '&', '*', '(', ')',  '=', '{', '}', '[',
+									 ']', '|', '\\', '<', '>', ',', '?', '/', ':', ';', '\'', '"', '~' };
+					if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+						tv.setKeyListener(new DigitsKeyListener(Locale.getDefault(), true, true) {
+							@Override
+							protected char[] getAcceptedChars()
+							{
+								return numbersAndPunctuationChars;
+							}
+
+							@Override
+							public int getInputType()
+							{
+								return InputType.TYPE_CLASS_TEXT;
+							}
+						});
+					} else {
+						tv.setKeyListener(new DigitsKeyListener() {
+							@Override
+							protected char[] getAcceptedChars()
+							{
+								return numbersAndPunctuationChars;
+							}
+
+							@Override
+							public int getInputType()
+							{
+								return InputType.TYPE_CLASS_TEXT;
+							}
+						});
+					}
+					customKeyListener = true;
 					break;
 				case KEYBOARD_URL:
 					Log.d(TAG, "Setting keyboard type URL-3", Log.DEBUG_MODE);
 					tv.setImeOptions(EditorInfo.IME_ACTION_GO);
 					textTypeAndClass |= InputType.TYPE_TEXT_VARIATION_URI;
 					break;
 				case KEYBOARD_DECIMAL_PAD:
-					textTypeAndClass = (InputType.TYPE_NUMBER_FLAG_DECIMAL | InputType.TYPE_NUMBER_FLAG_SIGNED);
+					// Overriding the inputType in order to match the shown keyboard on iOS (only numpad)
+					final int decimalInputType = INPUT_TYPE_CLASS_NUMBER | InputType.TYPE_NUMBER_FLAG_DECIMAL
+												 | InputType.TYPE_NUMBER_FLAG_SIGNED;
+
+					if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+						tv.setKeyListener(new DigitsKeyListener(Locale.getDefault(), true, true) {
+							@Override
+							public int getInputType()
+							{
+								return decimalInputType;
+							}
+						});
+					} else {
+						// For API levels lower than 26 we override the filter method of the DigitsKeyListener too
```

Comment:
We need to comment that we've copied Google's code; which is how we're back-porting it.
https://github.com/aosp-mirror/platform_frameworks_base/blob/master/core/java/android/text/method/DigitsKeyListener.java